users/toykeeper/toykeeper.h

```diff
@@ -0,0 +1,170 @@
+//#define DONT_USE_EEPROM
+
+
+// layers
+enum {
+    L_QWERTY = 0,
+    L_DVORAK,
+    L_UNDVORAK,
+    L_COLEMAK,
+    #ifdef USE_NUMPAD
+    L_NUMPAD,
+    #endif
+    #ifdef HAS_F_ROW
+    L_FLCK,   // toggle F1-F12 behavior (only used on 75% boards and bigger)
+    #endif
+    #ifdef USE_PALM  // palm Fn keys
+    L_LPALM,  // left palm key
+    L_RPALM,  // right palm key
+    L_BOTH,   // both palm keys
+    #else  // regular Fn keys
+    L_MOUSE,  // lower right
+    L_FN1,    // lower left
+    L_FN2,    // upper right
+    #endif
+    #ifdef HAS_XTRA_COLUMN
+    L_XC1,
+    L_XC2,
+    L_XC3,
+    #endif
+};
+#define L_BOTTOM (L_COLEMAK+1)  // lowest non-default layer
+
+enum custom_keycodes {
+    TK_INFO = SAFE_RANGE,  // print build version / info
+    // activate base layers like DF(N), and save them to the current switch slot
+    TK_DF_0,  // qwerty
+    TK_DF_1,  // dvorak
+    TK_DF_2,  // undvorak
+    TK_DF_3,  // colemak-dh traditional w/ TK mods
+    // like base layers, but for the extra keys only
+    TK_XC_0,  // Xtra Column mode 0
+    TK_XC_1,  // Xtra Column mode 1
+    TK_XC_2,  // Xtra Column mode 2
+    TK_XC_3,  // Xtra Column mode 3
+    // other custom keys
+    TK_IUUI,  // toggle U/I swap on dvorak layer
+    TK_ANY,   // the "Any" key, spits out random characters
+};
+
+// custom keys which don't need a custom enum
+#ifdef USE_PALM
+// Keyboardio Model-01
+#define TK_LPLM  MO(L_LPALM)
+#define TK_RPLM  MO(L_RPALM)
+#define TK_BOTH  MO(L_BOTH)
+#define TK_MOON  TK_LPLM
+#define TK_STAR  TK_RPLM
+#define TK_VOID  TK_BOTH
+#define TK_LGUI  MT(MOD_LGUI,_______)  // LGUI / oneshot(LALT) / hold(LALT)
+#define TK_RGUI  MT(MOD_RGUI,_______)  // RGUI / oneshot(RALT) / hold(RALT)
+#else
+// 65% to 75% layouts
+#define TK_MOON  MT(L_FN1,L_FN1)  // hold / lock / clear L_FN1
+#define TK_STAR  LT(L_FN2,KC_HOME)
+#define TK_MOUS  TT(L_MOUSE)
+#define TK_FN1   TK_MOON
+#define TK_FN2   TK_STAR
+#define TK_HOME  TK_STAR
+#ifdef OLD_QMK  // correct method didn't work until 2022-11-27 / QMK 19.0
+#define TK_LGUI  LGUI_T(KC_1)          // LGUI / left click
+#else
+#define TK_LGUI  LGUI_T(KC_BTN1)       // LGUI / left click
+#endif
+#define TK_RGUI  MT(MOD_RGUI,_______)  // RGUI / MouseLock
+#endif
+
+#define TK_LCTL  LCTL_T(TK_WARP)       // LCTL / Warp
+#define TK_RCTL  MT(MOD_RCTL,_______)  // RCTL / toggle(L_NUMPAD)
+#define TK_LALT  LALT_T(KC_BSPC)       // LALT / Backspace
+#define TK_RALT  RALT_T(TK_WARP)       // RALT / Warp
+#define TK_SESC  S(KC_ESC)
+#define TK_SINS  S(KC_INS)
+#define TK_WARP  KC_F23  // mouse warp, handled by host OS
+#define TK_BSTG  BS_TOGG
+#ifndef BS_TOGG
+#define BS_TOGG  _______  // don't break build on old versions
+#endif
+// TODO, maybe, someday: lockable control keys
+#define TK_LCLK KC_LCTL
+#define TK_RCLK KC_RCTL
+
+
+#ifndef DONT_USE_EEPROM
+// remember the user's preferred settings
+typedef union {
+    uint32_t raw;
+    struct {
+        #ifdef HAS_DIPSWITCH
```

Comment:
I'd honestly recommend removing the preprocessors here, so that if/when that this is saved into eeprom, the values aren't moved around and are consistent.

Reply:
I'm not sure I understand.  Are you saying that, with only 32 bits of user config space, it should allocate some of that space for dipswitch settings even on keyboards which don't have a dipswitch?

Reply:
The issue is when you change the state of the enabled features.  What happens if you have a board that has one of theses, but you flash it with it disabled. Then flash with it enabled?  You're gonna have a messed up value, since the bits are no longer in the same location. 

As for userspace config, it supports more than 32 bits now. You can use `EECONFIG_USER_DATA_SIZE` and that can get a lot more space. Though requires slightly different calls. You can see that added:
https://github.com/qmk/qmk_firmware/commit/f82e90bdeaab1b2ec5c96716731f0a4c63193de9